doc/architectural/memory-bounds-checking.md

```diff
@@ -0,0 +1,85 @@
+\ingroup module_hidden
+\page memory-bounds-checking Memory Bounds Checking
+
+cbmc provides means to verify that pointers are within the bounds of (statically
+or dynamically) allocated memory blocks. When the option `--pointer-check` is
+used, cbmc checks the safety of each pointer dereference in the program.
+Similary, the primitive `__CPROVER_r_ok(pointer, size)` returns true when it is
+safe to read from the memory segment starting at `pointer` and extending for
+`size` bytes. `__CPROVER_w_ok(pointer, size)` indicates if writing to the given
+segment is safe. At present, both primitives behave the same.
+
+Each memory segment is referred to internally in cbmc via an object id. Pointers
+are represented as bitvectors (typically of length 32 or 64). The topmost `n`
+bits encode the id of the memory segment the pointer is pointing to, and the
+remaining bits encode the offset within the segment. The number of bits used to
+represent the object id can be specified via `--object-bits n`. Object offsets
+are signed. This allows to distinguish the case of when a pointer has been
+incremented too much from the case of when a pointer has been decremented too
+much. In the latter case, a negative value would be visible for the offset
+portion of a pointer in an error trace.
+
+In the following, we describe how the memory bounds checks are implemented in
+cbmc for dynamically allocated memory. Dynamic memory is allocated in C via the
+`malloc()` library function. Its model in cbmc looks as follows (see
+`src/ansi-c/library/stdlib.c`, details not related to memory bounds checking are
+left off):
+
+```C
+inline void *malloc(__CPROVER_size_t malloc_size)
+{
+  void *malloc_res;
+  malloc_res = __CPROVER_allocate(malloc_size, 0);
+
+  __CPROVER_bool record_malloc = __VERIFIER_nondet___CPROVER_bool();
+  __CPROVER_malloc_object = record_malloc ? malloc_res : __CPROVER_malloc_object;
+  __CPROVER_malloc_size = record_malloc ? malloc_size : __CPROVER_malloc_size;
+
+  return malloc_res;
+}
+```
+
+Both internal variables `__CPROVER_malloc_object` and `__CPROVER_malloc_size`
+are initialised to 0 in the `__CPROVER_initialize()` function of a goto program.
```

Comment:
I think at some point it was decided that we stick to the Oxford spelling in documentation, i.e. "are initialized ..".